### PR TITLE
Add "finished" and "on hold" indicator to Mosaic browser

### LINKED ISSUE #7902: FR: [Finished mark on book's cover for mosaic display]
Hi, I've used Koreader for a year, and I found that the mosaic display has one disadvantage: when I browse in my library, I can't find the different between the books I have finished and the books I'm reading. Can we add a feature to add a "read" mark on book cover like Kindle does, or a "finished" line below like Kobo does?

### PR DESCRIPTION
**Pull request by melyux**
Closes #7902. Modifies the dogear icon we use for 'reading' books to change to reflect the other possible statuses (finished and on hold). Since the status was already being grabbed for each book (whether from cache or not), this simply uses that existing information. Additionally,

- Always show dogear indicator in file browser even with progress bar (which can still be toggled on and off); the option to turn off the dogear for History and Favorites still exists and works
- Don't show progress bar if book is completed
- Progress bar lighter but still visible if book is on hold
- Slightly lighter book outline to be less jarring (similar to the colors used by Kindle, Kobo, Book, etc. do it)

Other tweaks for some congruence:
- Larger dogear since the old one was too small
- 1 unit taller progress bar
- Move keyboard shortcut icon to top left to make room for progress bar

Fixes:
- Misspelling of "complete" status text led to the previous workaround not working ('show progress bar as 100% if book is finished')

Couldn't think of a better idea for an "on hold" indicator, so this is whatevs

**Before (reading, on hold, finished):**
before dogears
before progress

**After (reading, on hold, finished):**
after

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9939)

### CODE REVIEW

**Comment by mergen3107** [comment]
This is so great! Thank you!!!

**Comment by NiLuJe** [comment]
Oh, duh', chucking an icon in the folded bits is such a simple and neat way of solving that conundrum!

Gotta have to see how the increase in size feels like on real devices, though ;).

**Comment by mergen3107** [comment]
I love it how you made paused books have the folder ear, and the finished one unfolded :D

**Comment by poire-z** [comment]
> * Slightly lighter book outline to be less jarring (similar to the colors used by Kindle, Kobo, Book, etc. do it)

Please don't go from black to gray for such minor cosmetics preferences: gray on some devices does cause some annoying animated flickering, so don't create some when there isn't already some gray around. Among 9 covers, there will probably be many level of grays and lots of annoyances, but when using FakeCovers (Mosaic with text covers, or Mosaic with covers with books without covers), we used to get only black (titles + border), there's no reason to get black title and some annoying grayish border. (And at that level of thinness, it must be hard to notice it's not black - except for people/devices like me who will with the annoying animation :)

About the other stuff:
I can't test this PR at the moment, to really see what we will get. From the screenshots and what I can gather:
- no real opinion about the indicator on the dogear, but why not. Can I assume that people like me who don't change any book status will get the dogear blank as before?
- Visually, I'm not super fond of a larger dogear (I understand it's needed for the indicator): we had it small so it doesn't cover much of the original cover (and of our FakeCover bottom text), while still being just noticable enough.
- About the progress bar, I see that you have it now use the full bottom width of the cover. I don't think we had discussed that idea or even thought about it in #8915 (not saying it's then bad :) - but somehow, having it smaller gives it a bit more "relief"/layering. When it takes the full width like on your "After" screenshots, it' feels part of the cover and less like some added UI decoration. Pinging @uroybd for opinions and testing this tweak to his work.
- (I don't use Mosaic with file browser, only for History without opening hint, so I may not be affected but the visual changes if people really prefer the new stuff)

**Comment by melyux** [comment]
Fake covers can keep their black borders, and the normal covers deserve any gray flickering they asked for... heh. Let me give it a try and see if it looks weird to have both color borders though

- Yes. Well, everyone uses 2/3 statuses without even trying, so if you don't use the "on hold" status, you'll keep the old dogear ("reading") until you finish, at which point it'll become that "read"  symbol, which is even less intrusive than the old dogear. Can't hate me for that
- I think the old dogear was kind of funnily small where it added noise to the cover without clearly marking anything until you squint. Most ebook-reading software vendors seem to take great license with how they cover up the covers in their icon browsers, which I didn't like at first, but seems like they have a good reason. Until you start the book, you get to look at the pretty cover without any hinting whatsoever, and after you start, you've already "broken in" the book anyway so you can handle some markings to show your progress and status. At least unlike them, we allow the option to disable the dogear entirely for purists who just like to look at their covers
- About the progress bar... technically I still don't have it full width ;) it only shows up until the dogear. It's almost exactly the same width as it was before, but instead of that weird empty space to the left of it that I assumed was left so the keyboard shortcut icon wouldn't cover it up (it was, wasn't it?), that empty space is now to the right and occupied by the dogear. The previous iteration was honestly a jarring UI thing for me. If I added that beginning part again, the progress bar would now be narrower than it was before for no reason... I'm down for any better design mockups. I myself am not entirely a fan of how I made it look either, but IANAD (I am not a designer) which I'm very sad about. But I will stick to saying that I think this new progress bar placement looks better and less arbitrary than the previous blank space before the progress bar

To be honest, I think my new design is not great. I'm only partial to my new "read" marker, which is the best of the bunch. I want someone to come in and make it look good

**Comment by melyux** [comment]
Working on a separate PR to set these statuses from the on hold file dialogs, to complete the circle. Running into some lexical scoping mysteries there though, fingers crossed

**Comment by melyux** [comment]
@poire-z Actually now that I looked, it won't be so simple to do black borders for fake covers and the lighter borders for everything else, because the icons are gray themselves (I don't think IconWidget lets us overlay colors onto the icons), so we'd need separate icons for fake covers and normal covers. And separate progress widget drawing logic to determine its colors. And logic to display "on hold" book progress bar color as black separately if we have a fake cover.... etc

Lookswise, the fake covers look just as good with the slightly lighter borders as they did when they were black, but your problem is with the devices like yours that don't do gray so good, right? Can you give it a try and see how bad it is? (P.S. how does your device handle deleted files in the History viewer, since they're (even if fake covers) rendered light gray?)

**Comment by Frenzie** [comment]
> P.S. how does your device handle deleted files in the History viewer, since they're (even if fake covers) rendered light gray?

Some people (not me) prefer to autoclean history of deleted files.

**Comment by poire-z** [comment]
Got to test it.
There's some strong visual/optical effect/illusion that makes the normal/reading dogear (on the left, below) bigger than the "finished" dogear (on the right) - so strong it feels like the developers were blind and didn't notice the icons were not the same size - but then, what else to expect from developpers, we all know they don't have any taste ?! :)

image
(it's more noticable on my device than on this screenshot.)

So, if anything, you could make the "reading" icon smaller (as it was?) - and just have the "finished" dogear larger (and no idea about "on hold" that would need to be smaller given its shape).

> (P.S. how does your device handle deleted files in the History viewer, since they're (even if fake covers) rendered light gray? Maybe this'd be served better with a universal "pure black and white" option to handle these devices...)

Well, they are gray - and rare as  I have the option set to remove deleted files from history, so I don't see them often.
I'm also fine with the other grays we may have, like in the bottom menu. The only thing I do change to black in my own patch is the gray separators in the top menu, because there, when switching menu, the flickering is noticable.

> Lookswise, the fake covers look just as good with the slightly lighter borders as they did when they were black, but your problem is with the devices like yours that don't do gray so good, right? Can you give it a try and see how bad it is? 

I did. And strangely, I didn't get to notice the refresh animation as much as I feared (because actually, when drawing the same gray over the same gray, my Kobo doesn't need to do this refresh/flickering).
But I still don't like these gray borders:
- they look thinner and more dirty (not as full and clean) than black on my Kobo (I think gray may also look as dirty with horizontal lines/separators, but it hurts a lot less with mostly ignorable separators, than with wrapping rectangles that make the main visible thingies)
- also (again, may be my Kobo only), there is more ghosting of past vertical cover edges: switching from a page with real covers (variable widths) to the next one with fake covers, and I get to see more of these past vertical edges than when using pure black.
- I don't find them particularly nicer than black borders, even after a full refresh to get everything painted nicely: for Fake covers, it doesn't feel justified: black is just right around black title text. And around real covers, I also don't feel they bring anything: on colored covers, I hardly notice them, whether they are black or gray - and on white covers, same feeling of thinness and lack of contrast that with fake covers.
- so, this choice feels like personal taste, and possible technical side effects advise against it :/

> Actually now that I looked, it won't be so simple to do black borders for fake covers and the lighter borders for everything else, because the icons are gray themselves

So, have these icons black :)

**Comment by mergen3107** [comment]
There has been a suggestion to replace the dogear with a ribbon because dogears promote bad attitude towards books :D so maybe the design’s gonna change…

**Comment by mergen3107** [comment]
Here: https://github.com/koreader/koreader/issues/9965

**Comment by NiLuJe** [comment]
Gentle ping ;).

**Comment by melyux** [comment]
Back. Have to redraw the icons and borders in black and post a screenie

**Comment by melyux** [comment]
Made everything black again, except the progress bar when a book is on hold.

**Comment by poire-z** [comment]
Thanks for the back to black - which is enough for me to not veto :)

Not super fond of the progress bar "merging" into the cover bottom, as expressed above, but I understand your points. And I won't see it, so no problem.

Still a bit overwhelmed by the large dogear. Not as shocked as when I tried it 5 weeks ago though, so I guess people can get used to it.
Any thought about the optical illusion of dogear size mismatch mentionned above ? (could lead to a compromise to make the "read" and "on hold" ones a bit smaller :) dunno how complicated this would get, to have different sizes though). Still not using this, so I won't mind.
But what somehow catch my eyes is when these dogear overlap with the bottom text (the filename) in FakeCovers. I don't know if I made sure they can't overlap when I made it, but I think i would have "fix" that if I had seen such overlaps. Dunno if this could be "fixed" by increasing padding/margin for the bottom text in FakeCover (possibly only if these hints are used, as no need to waste width/space when there's no risk of things overlapping).
And also the progress bar feels overlapping (or just about to overlap, the bottom of "p" or "g" when a not-small font size gets used touch the progressbar).

**Comment by melyux** [comment]
@poire-z It wouldn't be bad to make the reading and on hold dogears smaller and keep the finished one the current size, but then the Pause icon in the On Hold dogear would be too small to see. It's already pretty small. We could try to come up with a better icon for "on hold", but I can't think of any... or add something into the "reading" dogear to avoid the optical illusion (maybe a percentage number somehow? or a "play" icon to match the pause icon? could be cool)

Let me see what I can do about the filename-dogear overlap (which already existed for me with the old dogear with some of my books). Or we could leave it as it is, like those magazine covers where the cover model's head blocks part of the magazine title but no one seems to care ;)

I'll move the filename up a bit in any case to compensate for the slightly taller progress bar

**Comment by melyux** [comment]
Side note. The dogear is kinda weird because it actually looks like you dogeared the entire book, all 400 pages of it

**Comment by melyux** [comment]
I don't see why we show the full filename at the bottom of a FakeCover anyway. Either the book has metadata for a title, in which case we don't need to show the filename (normal covers don't show the filename, so why should a FakeCover when we have metadata instead?), or if there's no metadata, we currently use the filename minus the extension as the title anyway, so no need to repeat the filename at the bottom: just show the extension/filetype ("PDF" or "EPUB") at the bottom. What do you think?

**Comment by poire-z** [comment]
> or add something into the "reading" dogear to avoid the optical illusion (maybe a percentage number somehow? or a "play" icon to match the pause icon? could be cool)

No, won't be cool to me :) Let it be as plain as it is, I will only see this one, I don't need it busy with a small icon.

> I don't see why we show the full filename at the bottom of a FakeCover anyway. Either the book has metadata for a title, in which case we don't need to show the filename (normal covers don't show the filename, so why should a FakeCover when we have metadata instead?), or if there's no metadata, we currently use the filename minus the extension as the title anyway, so no need to repeat the filename at the bottom: just show the extension/filetype ("PDF" or "EPUB") at the bottom. What do you think?

I think no :)
Not all people (and not me) tidy their metadata, and we can get crap. The filename (which is the only thing I can edit and tidy) gives me the info that the metadata may hide. Ie.:

Duplicated/hacked/fixed same book:
image

Limited metadata (worth it for a first eye scan), but the filename gives more:
image image

I added the filename initially just to add something so we can get the book title centered (it would usually be the publisher company name at the bottom in a real cover), but it ends up useful (at least to me, and I would miss it).
If you wanted to replace it with some Series info, this should be an option (default or not, as long as it's disableable).

**Comment by Frenzie** [comment]
> just show the extension/filetype ("PDF" or "EPUB") at the bottom. What do you think?

Hard pass from me, metadata is junk.

**Comment by Frenzie** [comment]
> just show the extension/filetype ("PDF" or "EPUB") at the bottom. What do you think?

Hard pass from me, metadata is junk. :-)

**Comment by melyux** [comment]
Implemented making the FakeCover bottom text un-overlappable by the dogear.

@poire-z Yeah, you're probably right. Tried the play icon and didn't like it. Leaving things as they are then

@Frenzie @poire-z Hands off the filename then :) If you're serious about being open to the option thing, I'll try adding that 4th option to the "Series" submenu ("Show series metadata instead of filename")

@Frenzie I get wanting to keep the filename even when metadata is present (metadata is junk), but for the suggestion of showing just the extension at the bottom ("epub"), that would have only applied when the metadata doesn't exist anyway. In that case right now, we show the exact same thing (the filename) for the title _and_ at the bottom of the FakeCover*.

*Well, we do do some transformations to the filename to make it prettier as a title

**Comment by Frenzie** [comment]
> but for the suggestion of showing just the extension at the bottom ("epub"), that would have only applied when the metadata doesn't exist anyway. 

In that case it doesn't really matter from that angle, though presumably making it different still adds maintenance burden and visual distraction.

**Comment by melyux** [comment]
> > but for the suggestion of showing just the extension at the bottom ("epub"), that would have only applied when the metadata doesn't exist anyway.
> 
> In that case it doesn't really matter from that angle, though presumably making it different still adds maintenance burden and visual distraction.

This looks fine, right? Almost a one-liner and using an existing util method

FileManager_2023-02-09_015654

**Comment by Frenzie** [comment]
In isolation, but you can clearly see how the ones that stand out are jumping around for your attention in context.

**Comment by melyux** [comment]
Seemed okay to me, I guess if your eye's looking for the filename on every item, it might be somewhat jarring to see just an extension on some of them. But that's the case for many of them already (see `kluit-web.svgz`, I've got a few like that lying around)

**Comment by Frenzie** [comment]
Oh, I didn't mean you should revert it, or at least not necessarily. I mainly wanted to be perfectly clear about what effect I was referring to.

**Comment by melyux** [comment]
Hmm, what if we had it be ".jpeg" instead of "JPEG". Might go some way towards negating the effect

**Comment by poire-z** [comment]
Feels a bit inconsistent to not always have a lengthy bottom text with the filename.
But not strictly opposed, it's a little hint that the title is actually ~equal to the filename and you don't need to force your eyes to read that small thing and find that out yourself :)

> Hmm, what if we had it be ".jpeg" instead of "JPEG". Not as pretty in isolation but might go some way towards negating the effect

Dunno, I prefer without the dot aesthetically-wise.
But (in French? or it might be just me :), I mostly say "c'est un point PDF", "c'est un point EPUB" ("it's a dot-PDF"), so... it would read ok :)

> If you're serious about being open to the option thing, I'll try adding that 4th option to the "Series" submenu ("Show series metadata instead of filename")

Nothing against that :) (not _seriously open_, but I'm neither joking nor closed :)

**Comment by NiLuJe** [comment]
> Dunno, I prefer without the dot aesthetically-wise.

Ditto ;).

I also kinda like the idea, regardless of potential inconsistencies, FWIW.

(Full disclosure, though: I'm not a FakeCover user, because I take care of my metadata ;p).

**Comment by Frenzie** [comment]
If it's done I concur that without the dot is better. But I like the status quo. :-P

**Comment by melyux** [comment]
We could have the new 4th setting also enable this change so that if you're showing filenames, we keep the old behavior, but if you set that 4th setting to deprioritize the filename, you see series data for files with metadata and "PDF" for files without metadata. So it'd add consistency for both options

**Comment by Frenzie** [comment]
Sorry, which new fourth setting are you referring to?

**Comment by melyux** [comment]
We have the 3 settings for where to add Series metadata: title, author, or on its own line. The 4th possible setting would be something like "Instead of filename", to show the series name at the bottom instead of the filename for books with metadata

So that setting could also control whether you see a repeated full filename at the bottom for books without metadata, or just the capitalized extension

**Comment by poire-z** [comment]
^ These 2 things (series, vs sole extension at the bottom when title==filename) are orthogonal: I may want the latter without the former :) so feels a bit odd to merge these 2 behaviours in a single setting named after only one behaviour.

I'd go with it, the spirit of this comment surpass the feeling of inconsistency :):
> But not strictly opposed, it's a little hint that the title is actually ~equal to the filename and you don't need to force your eyes to read that small thing and find that out yourself :)

With your 4th setting, you would still show the filename (or extension if you go with that) when there is no Serie metadata ? or would you have it blank to make it clear there is no Serie metadata ?

**Comment by melyux** [comment]
> ^ These 2 things (series, vs sole extension at the bottom when title==filename) are orthogonal: I may want the latter without the former :) so feels a bit odd to merge these 2 behaviours in a single setting named after only one behaviour.

Right? But the only-the-extension-at-the-bottom thing doesn't seem different enough to warrant its own setting, and the setting doesn't have to be named for series; it could be something akin to "I don't want to see filenames" 😇 

> With your 4th setting, you would still show the filename (or extension if you go with that) when there is no Serie metadata ? or would you have it blank to make it clear there is no Serie metadata ?

I'd leave it blank for clarity if there's no series metadata

**Comment by melyux** [comment]
Anyway, aside from the possible FakeCover improvements we can continue with, the original change is ready I think. I fixed the FakeCover bottom line's progress bar overlap and dogear overlaps

**Comment by poire-z** [comment]
Tested again, and ok for the no-longer text-overlapping bottom stuff.
(Stilll overwhelmed by the large dogear but ok :)
(Also, regarding the optical illusion, if you can't make the "read" and "on hold" smaller, you could make the "finished" larger - but well, not liking "larger", I can't really suggest that :)

On my kobo, the dogear lines are thicker than the book cover borders (unlike on your screenshots), which feels a bit odd, but it could somehow be justified by the fact that the dogear dogearing frontward, it is closer to the eyes of the reader, so appears larger and thicker :)
I checked that it was already that thicker before the PR, so not really something new (it was less stricking as the dogear was smaller).

Also, may be a tad more bothering, the progress bar stops at the dogear, like in your screenshots:
image
When you see this, you can have some doubts the progress bar stops at the dogear, or continues below it up to the right border of the cover - and if the dogear has dogeared that tail end of progress bar too :). So, you can have doubts when trying to appreciate the % from that progress bar, not being sure how long it really is. On that screenshot, is that > 50% or < 50% ? And when it's near the end, are you near the end of the book, or only at 80%?

Anywhow, good with me.

**Comment by NiLuJe** [comment]
> Also, may be a tad more bothering, the progress bar stops at the dogear, like in your screenshots:

I concur with that feeling, especially since we can't see the bar's right border ;).

**Comment by melyux** [comment]
We could try putting the progress bar at the top, full width, see how that looks. Or what might be better is to put the dogear at the top right instead of at the bottom and making the progress bar full width at the bottom

**Comment by poire-z** [comment]
> Or what might be better is to put the dogear at the top right instead of at the bottom

I won't let you redecorate our whole living room ! You can change the carpet, but let the TV where it is ! :)
We're used to that, it's at the bottom right in Detailed list mode, don't change this because of my comment, let's just forget my comment :)

Or... just make the progress bar a tad smaller ?
I see:
`progress_widget.width = cover_item.width - corner_mark_size + 2`
Looks like you added the +2 just for that, no ?
And it might 1) not be enough, possibly because 2) you did not Screen:scaleBySize() it, as you should :)

**Comment by melyux** [comment]
Alright alright but it could have solved everything ;( Give love a chance

Actually that + 2 was to make sure the progress bar aligned without a gap to the dogear, previously the progress bar was slightly off kilter.

Dogear you want to stay at the bottom, but what about progress bar at the top? Reasoning being that making the progress bar smaller would add a visually annoying gap. I'll see how the progress bar looks with thicker borders too per @NiLuJe

**Comment by NiLuJe** [comment]
The border experiment seems to confirm that there might be a scaling issue, as it reveals a pixel or two of overlap ;).

That's my main concern with it, I'd be much happier actually *seeing* the bar's border, separate from the dogear (or right next to it, but not the two merging together).

I'd go so far as to say the same thing about the *left* border, but that's less visually confusing there ;).

So, by default, I prefer a shorter bar because it "fixes" that, but I'm not necessarily set on the idea of a shorter bar, really, I just want to see the separation between the bar & the dogear ;).

**Comment by NiLuJe** [comment]
Oops, forgot the zoom-in screenshot :D
overlap

**Comment by melyux** [comment]
And here's dogear at top right... just as an experiment:

dogear top

**Comment by poire-z** [comment]
> Smaller bar:

I prefer this one to all the others.

> And here's dogear at top right... just as an experiment:

There, it feels too like our regular book page bookmark, possibly hinting that the book does have bookmarks - so, confusing.

**Comment by NiLuJe** [comment]
I'm less offended by the latest upside down proposition than the others, but, yeah, still prefer a shorter bar in the usual layout ;p.

**Comment by melyux** [comment]
You guys seem to agree, so I put the old progress bar back. (I will mark a point of protest though, I think this looks really ugly and getting rid of this floating progress bar was part of the inspiration for this PR)

shorter bar 2

**Comment by poire-z** [comment]
You don't have to keep it as small as it was.
It looks better with it a bit larger like on your "Smaller bar" screenshot. You just need a small gap before the dogear so we can notice that it ends there.

**Comment by melyux** [comment]
I think that one looks even worse than this, so if it doesn't matter, this one's slightly better. Maybe centering the bar will help, let me try that

**Comment by NiLuJe** [comment]
> I'm not necessarily set on the idea of a shorter bar, really, I just want to see the separation between the bar & the dogear ;).

^ ;).

**Comment by poire-z** [comment]
Another option could be to have the progress bar below the cover, the size, so there is no doubt that it doesn't go over the dogear area - and it would not overlap with any part of the cover. Dunno how it would look.
image

But it would probably complexify the refresh areas, and may be overlap with the cover below. I wouldn't really want a cover item solution to need to tweak the whole 3x3 cover layout.

**Comment by melyux** [comment]
Looks like this now:
shorter prog bar 3

I still think the gap looks random and dirty, but I can always patch it

@poire-z I thought about that and mocked it up and yeah it doesn't looks so good eh? Ah well

**Comment by mergen3107** [comment]
Here is the idea: have the same length offset on both sides of the bar, like 5% of the bar width. 

So the bottom will look like offset-bar-offset-dogear

**Comment by mergen3107** [comment]
Plus add the bottom offset of the bar, so that it is vertically center-aligned w.r.t. the dogear height.

**Comment by Frenzie** [comment]
I think that's worth a try. :-)

I think this one for example looks pretty good. What was the concern, that it's not completely clear if it continues underneath the dogear?

**Comment by poire-z** [comment]
> What was the concern, that it's not completely clear if it continues underneath the dogear?

Yes, https://github.com/koreader/koreader/pull/9939#issuecomment-1425450621

**Comment by NiLuJe** [comment]
Another thing to check that's horribly broken on master: if you disable "hints" (i.e., the dogear), but enable the progress bar, it appears to still account for the dogear or something, and looks... very weird ;).

(Haven't yet checked with this applied ;)).

**Comment by melyux** [comment]
@mergen3107 Do you mean the bar is floating above the bottom of the book in your idea? I'm trying to recreate it

@NiLuJe The dogear and the progress bar seem to be mutually exclusive in master, unless my code base is screwed up (probable)

**Comment by NiLuJe** [comment]
> @NiLuJe The dogear and the progress bar seem to be mutually exclusive in master, unless my code base is screwed up (probable)

Yup, that made the issues with the progress bar placement even more obvious, which you discussed in your original message.

(I had edited the above message to clarify that, but apparently GH ate the edits -_-").

**Comment by mergen3107** [comment]
@melyux 
Yes, floating slightly above the bottom.

**Comment by melyux** [comment]
That will be a problem for our FakeCover enjoyers by covering up the bottom text. I'll try it anyway, and a non-offset version. I tried the non-offset version earlier but didn't like it for some reason and didn't even bother posting a screeny of it, but new day new tastes

**Comment by poire-z** [comment]
> Current state of the branch, with a gap between the end of the progress bar and the dogear

If you're happy enough with this, should we merge ? and leave lucky tweaks for another PR?
As mentionned in your other PR, I'd be happy with you suggesting the final message for the squashed commit :)

**Comment by melyux** [comment]
Let me present these for your consideration before we merge. Implementing @mergen3107's ideas, and I think they look pretty good actually. Especially the offset one, it looks quite professional. Might even be worth pushing the bottom FakeCover text up a bit...

Offset:
prog bar floating middle

Non-offset:
bar middle 1

**Comment by poire-z** [comment]
I prefer the second one, stuck to the bottom (and centered in the remaining area looks nice.

> Especially the offset one, it looks quite professional if you don't mind losing some cover details.

"Professional" as in: let's mask 1/8th of the cover so it looks cool and centered :)
And you're going to make FakeCover unbalanced just for that (even those not opened)?
I prefer developper sobriety to marketing fancy :)

**Comment by poire-z** [comment]
But anyway, I'm not the customer for this feature, so it's not for me to say what to sacrifice for fancyness, I just wish it does not complexify the othercode (ie. FakeCover) so we don't have to text N x M x P combination of settings anytime we had to tweak something.

@uroybd : you added the progressbar to Mosaic mode items, any thought on all this ?

**Comment by mergen3107** [comment]
Offset one looks really nice! Didn’t expect it to :D

**Comment by melyux** [comment]
Here's FakeCover compensation for the floating bar. Not _too_ crazy, plus we gain from not compensating for the dogear anymore. If we don't go with it, I'll probably end up using it as a personal patch anyway

floating prog bar fakecover

**Comment by mergen3107** [comment]
If the filename is two-line, will the second line be hidden underneath the dogear? Or the whole two-line block moves upward?

**Comment by poire-z** [comment]
Can we see this FakeCover near a non-opened FakeCover, near an opened & non-opened RealCover?

**Comment by melyux** [comment]
@mergen3107 The whole block would move upwards if the book has a progress bar

**Comment by NiLuJe** [comment]
I'm not necessarily a huge fan of how it looks on a real cover, but I do prefer the offset one there to the non-offset (at least with a small bar like that, thought might change with a longer bar).

It does look pretty nice on a FakeCover, though.

Like @poire-z said, I'm not going to see a lot of those anyway, so, if offset looks like the best compromise so far, I'm cool with it.

**Comment by NiLuJe** [comment]
The fact that the real cover in question has an extra black border on the bottom might be throwing my perception ther,e though ;).

**Comment by melyux** [comment]
Alright, this is pretty exciting. Here's

Real cover with both progress bar and hints on:
real cover, prog 1, hint 1

Real cover with progress bar on and hints off:
real cover, prog 1, hint 0

Fake cover with both progress bar and hints off:
fake cover, prog 0, hint 0

Fake cover with progress bar off and hints on:
fake cover, prog 0, hint 1

Fake cover with progress bar on and hints off:
fake cover, prog 1, hint 0

Fake cover with both progress bar and hints on:
fake cover, prog 1, hint 1

**Comment by melyux** [comment]
@poire-z Ignore the nonsensical arrangement

example

**Comment by NiLuJe** [comment]
That could work, yeah ;).

**Comment by poire-z** [comment]
Well, if it is growing on @NiLuJe , I'll just shut up :) Go ahread :)

**Review by NiLuJe — APPROVED** [review]
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9939)* status: all files reviewed, 1 unresolved discussion (waiting on @melyux)

___
*[`plugins/coverbrowser.koplugin/mosaicmenu.lua` line 583 at r1](https://reviewable.io/reviews/koreader/koreader/9939#-NOH1t3H7sCuyqozatVo:-NOH1t3H7sCuyqozatVp:bo6evp1) ([raw file](https://github.com/koreader/koreader/blob/d4ea57807403ba42fd9de0d0bbedf33d607245e0/plugins/coverbrowser.koplugin/mosaicmenu.lua#L583)):*
> ```Lua
>                     pages, percent_finished, status = unpack(self.menu.cover_info_cache[self.filepath]) -- luacheck: no unused
>                 else
>                     local docinfo = DocSettings:open(self.filepath)
> ```

re: #10119

File: plugins/coverbrowser.koplugin/mosaicmenu.lua
Comment (poire-z) [inline plugins/coverbrowser.koplugin/mosaicmenu.lua:793]: Just asking (dunno the answer): is it alright to use here `Screen:scaleBySize(16)` (probably some experimental value you found nice :), and not some fraction of `cover_item.width` or `corner_mark_size` ?
Screen:scaleBySize() will scale according to the screen size.
It may end up that all other things also end up being a fraction of the screen size, so we end up with all things sized related to the screen size, but I dunno :)

File: plugins/coverbrowser.koplugin/mosaicmenu.lua
Comment (poire-z) [inline plugins/coverbrowser.koplugin/mosaicmenu.lua:916]: (May be if over users find the corner mark overwhelming, we could go with 12 as I seem to have found it alright :)

**Review by poire-z — APPROVED** [review]

File: plugins/coverbrowser.koplugin/mosaicmenu.lua
Comment (melyux) [reply plugins/coverbrowser.koplugin/mosaicmenu.lua:793]: It's probably related to the corner mark size, let me draw up some golden ratio charts and whatnot and figure out the fraction

File: plugins/coverbrowser.koplugin/mosaicmenu.lua
Comment (melyux) [reply plugins/coverbrowser.koplugin/mosaicmenu.lua:916]: I tried 12 but found it too small, but we can always reduce if the people rebel. Otherwise we may have to chip in for an early eye appointment for you and fix the problem from that front 🧐

File: plugins/coverbrowser.koplugin/mosaicmenu.lua
Comment (melyux) [reply plugins/coverbrowser.koplugin/mosaicmenu.lua]: Done, it is indeed a derivative of `corner_mark_size`.

**Review by NiLuJe — APPROVED** [review]
Haven't looked too much into it, but the updated maths sounds sensible ;).

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/koreader/koreader/9939#-:-NOQ0sxR02Q390TYZX6k:b95q5hx)* status: all files reviewed, 3 unresolved discussions (waiting on @melyux and @poire-z)

**Comment by poire-z** [comment]
"encore!" for the _final message for the squashed commit_ :)

**Comment by melyux** [comment]
Simply "Mosaic: Book status, floating progress bar"?

**Comment by poire-z** [comment]
That's a good commit title. May have some words in the commit details about the new dogear icons ?

**Comment by melyux** [comment]
```
- New dogear icons in Mosaic cover view to indicate each possible book status: 'Reading', 'On hold', 'Finished'
- Progress bar redesigned to be floating, taller, thicker bordered in Mosaic cover view
- FakeCover bottom text (filename) adjusted to not overlap with progress bar and dogear if they exist
- Mosaic shortcut letter moved from bottom left to top left
```

**Comment by mergen3107** [comment]
@melyux 
This looks absolutely gorgeous! Thank you very much!

Have a look:

Reader_2023-02-18_203443

Now that History (and Favorites) pages look so neat and professional, I would suggest makes an option to force the same size for all covers.
So, it would something like this (when I did `Ignore cover` via long tap on each book here), but with the actual covers stretched:
Reader_2023-02-18_203258

Another suggestion is related to the FM setting:
Top menu - Settings - Mosaic and detailed list settings - 
here it says `Show progress % in mosaic mode`.

With current option, this should be changed to `Show book progress bar in mosaic mode` or something like that.

Otherwise, how about adding (after a 55124 day debate on design, I guess :D) the actual option to add the progress percentage to the progress bar?
So there would be these both options:
`Show progress % in mosaic mode`
`Show book progress bar in mosaic mode`

This is not critical at all, just wanted to bring attention to this new inconsistency.

**Comment by NiLuJe** [comment]
> add the progress percentage to the progress bar?

That's currently not a ProgressWidget at all, and my meager experience doing that in a much easier context (bitmap fonts) was... trickier than one would expect ;).

**Comment by mergen3107** [comment]
Not necessarily inside the bar, maybe right before?

E.g.
`87% progressbar dogear`

**Pull request merged**
